ext/bcmath/libbcmath/src/recmul.c

```diff
@@ -261,47 +347,492 @@ static void bc_standard_mul(bc_num n1, size_t n1len, bc_num n2, size_t n2len, bc
 		 * overflow, so digit adjustment is performed beforehand.
 		 */
 		if (UNEXPECTED(count >= BC_MUL_MAX_ADD_COUNT)) {
-			bc_digits_adjustment(prod_uint, prod_arr_size);
+			bc_digits_adjustment(prod_int, prod_arr_size);
 			count = 0;
 		}
 		count++;
 		for (size_t j = 0; j < n2_arr_size; j++) {
-			prod_uint[i + j] += n1_uint[i] * n2_uint[j];
+			prod_int[i + j] += n1_int[i] * n2_int[j];
 		}
 	}
 
 	/*
 	 * Move a value exceeding 4/8 digits by carrying to the next digit.
 	 * However, the last digit does nothing.
 	 */
-	bc_digits_adjustment(prod_uint, prod_arr_size);
+	bc_digits_adjustment(prod_int, prod_arr_size);
 
 	/* Convert to bc_num */
-	*prod = bc_new_num_nonzeroed(prodlen, 0);
-	char *pptr = (*prod)->n_value;
-	char *pend = pptr + prodlen - 1;
+	bc_mul_convert_int_to_bcd(prod_int, prodlen, prod_arr_size, prod);
+
+	efree(buf);
+}
+
+/*
+ * In divide-and-conquer calculations, determine whether the calculation length is
+ * such that digits should be adjusted to prevent overflow during calculation.
+ * Digit adjustment is performed when the calculation length is a power of
+ * BC_REC_MUL_DO_ADJUST_EXPO.
+ */
+static inline bool bc_rec_mul_near_overflow(size_t calc_arr_size)
+{
+	if (EXPECTED(calc_arr_size < BC_REC_MUL_DO_ADJUST_EXPO)) {
+		return false;
+	}
+
+	while (calc_arr_size > 0) {
+		calc_arr_size /= BC_REC_MUL_DO_ADJUST_EXPO;
+		if (UNEXPECTED(calc_arr_size == 1)) {
+			return true;
+		}
+	}
+	return false;
+}
```

Comment:
I don't understand why calculations can overflow, according to Wikipedia there is a way to *always* do the required steps without multiplication overflow.

See the last paragraph of https://en.wikipedia.org/w/index.php?title=Karatsuba_algorithm&oldid=1190009898#Implementation